feat(ui): new focus manager

milesap opened the pull request:
Notable changes
- $rootElement must have an 'rv-trap-focus` property
- Any element located outside the $rootElement must have `rv-trap-focus`
  property set with a value of the $rootElements id
- The jquery methods focus, preventDefault, stopPropagation, and
  stopImmediatePropagation cannot be called on viewer elements - this is
  strictly disabled by the manager (see documentation for how to access
  these methods). Elements not a part of the viewer are not affected

Closes #1210, #1213, #1202, #1208, #1205

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/1293)

<!-- Reviewable:end -->

alyec commented:
General notes from the code review:
- add support for arrow keys
- move focus manager outside the viewer application
- add an `RV` option to disable the focus manager
- add public calls for `mdDialog` and `asyncApply`; replace `$rootElement`

---

Reviewed 37 of 37 files at r1.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[src/app/layout/focus.service.js, line 40 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV5hkdd9dDv5VmITaiE:-KV5hkdd9dDv5VmITaiF:bf2cwxw) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L40)):*
> ```JavaScript
>         const VIEWER_SELECTOR = `[rv-trap-focus=""], [rv-trap-focus="${$rootElement.attr('id')}"]`;
> 
>         let history = []; // ordered list of elements which has received focus
> ```

I don't think I saw history being assigned to in the code review, if that's the case can you change this to const?

---

*[src/app/layout/focus.service.js, line 41 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV5hcC4jPrO5UIGcS7C:-KV5hcC4jPrO5UIGcS7D:b-4x6miz) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L41)):*
> ```JavaScript
> 
>         let history = []; // ordered list of elements which has received focus
>         let linkedList = []; // list of element pairs which focus moves between
> ```

Can you expand the comment to describe the structure of the elements?  Also, is this possible to const?

---

*[src/app/layout/focus.service.js, line 191 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV5iQrYWMYgdJGPu5SS:-KV5iQrYWMYgdJGPu5ST:b-ymeych) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L191)):*
> ```JavaScript
> 
>             return $q(resolve => {
>                 $rootScope.$applyAsync(() => {
> ```

add comment about `applyAsync` and tooltips (also convert this to an API call when this moves out of the viewer code)

---

*[src/app/layout/focus.service.js, line 288 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV5jk6ZSHEU4h7kOevr:-KV5jk6ZSHEU4h7kOevs:b-766s9d) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L288)):*
> ```JavaScript
>          * @return    {Object}    a jQuery element which is next/previous to the reference element provided
>          */
>         function focusableSearch(element, forward, gotoEnd = false) {
> ```

add examples in jsdoc: @james-rae can make some fancy mspaint diagrams

---

*[src/app/layout/focus.service.js, line 296 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV5jZ6yMfj3Os0GjjWb:-KV5jZ6yMfj3Os0GjjWc:b-4jttz6) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L296)):*
> ```JavaScript
>             let refElem = element;
>             // only stores actual elements - is not empty
>             let lastKnownElement = gotoEnd ? foundElement : $();
> ```

since `lastKnownElement` can be a set rename to something like `lastElementSet`

---

*[src/app/layout/focus.service.js, line 307 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV5jFkd-QkU446mRWna:-KV5jFkd-QkU446mRWnb:b-k2qaph) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L307)):*
> ```JavaScript
>                     // no elements have been found, so lets find the first/last focusable in focus trap to loop to
>                     } else {
>                         return focusableSearch(refElem, !forward, true);
> ```

add a comment about how this handles "bouncing"

---

*[src/app/layout/focus.service.js, line 379 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV5k6hycGvCS94on8p7:-KV5k6hycGvCS94on8p8:b-5qgbhs) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L379)):*
> ```JavaScript
>          */
>         function hasLink(forward) {
>             const lastHistoryElem = $(history.slice().reverse().find(el => el.is(elemIsFocusable)));
> ```

replace with call to `lastVisibleHistoryElement`

---

*[src/app/layout/focus.service.js, line 405 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV60bA0hLxupt7NUoO0:-KV60bA0hLxupt7NUoO1:b-adybyb) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L405)):*
> ```JavaScript
>             // no link exists
>             } else {
>                 const focSearch = dir => focusableSearch($(document.activeElement), dir);
> ```

this function can be inlined into the `else` block below

---

*[src/app/layout/focus.service.js, line 430 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV62oKMcZDnVLVBfuz0:-KV62oKNSpgTqDXEtaRb:biuop0m) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L430)):*
> ```JavaScript
>             setStatus(statuses.WAITING);
> 
>             $mdDialog.show({
> ```

translations needed (use substitution strings for the HTML)

---

*[src/app/layout/shell.directive.js, line 78 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV63Al_5OKsvdL-TUdJ:-KV63Al_5OKsvdL-TUdK:b1b1ijz) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/shell.directive.js#L78)):*
> ```JavaScript
>                 }
>             },
>             {
> ```

create a `dialogService` which can wrap `mdDialog.open`

---

*[src/app/ui/filters/filters-default-menu.directive.js, line 47 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV63wuO0sxY7hWR99wg:-KV63wuO0sxY7hWR99wh:bbmu2i) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/ui/filters/filters-default-menu.directive.js#L47)):*
> ```JavaScript
>         self.dataPrint = dataPrint;
>         self.dataExportCSV = dataExportCSV;
>         self.appSelector = $rootElement.attr('id');
> ```

there should be an appInfo service which provides the `id` directly instead of looking up every time

---

*[src/app/ui/loader/loader-menu.directive.js, line 37 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV64_GRjHlT273UHGk5:-KV64_GRjHlT273UHGk6:b-fv2dl7) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/ui/loader/loader-menu.directive.js#L37)):*
> ```JavaScript
>         // TODO: need a better way to determine if the layer loader is active or not
>         self.state = stateManager.state;
>         self.appSelector = $rootElement.attr('id');
> ```

appInfo?

---

*[src/content/styles/layout/_main.scss, line 55 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV64lxYLsvsK6IA_I_e:-KV64lxYLsvsK6IA_I_f:b-5b6u4q) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/content/styles/layout/_main.scss#L55)):*
> ```SCSS
> 
> div.md-button {
>     line-height: 20px;
> ```

add a comment about what this does

---

*[src/content/styles/modules/_toc.scss, line 320 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV64rbqBq6N9n-lwepm:-KV64rbqBq6N9n-lwepn:b8231te) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/61daf32952e4b1772429b35d47e3d8cf48091bf7/src/content/styles/modules/_toc.scss#L320)):*
> ```SCSS
>         }
> 
>         // hide layer toggles except the visibility toggle
> ```

add this back and figure out a good way to handle this during the next UI meeting

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-:-KV5h9mlwOvf4GtGhgtv:b-c2271v)*
<!-- Sent from Reviewable.io -->

milesap commented:
Updated estimate to 5 to complete changes as discussed in the focus meetings. Largest portion of the work will be moving the code from being an angular service to a non-angular page "plug-in".

milesap commented:
- Will add translation for dialog
- Will rebase on v1.4.0 target (currently develop)

---

Review status: 10 of 40 files reviewed at latest revision, 19 unresolved discussions.

---

*[src/app/common/router/displaymanager.service.js, line 110 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV-r4_vwBufe64J1BxL:-KV0CRIsIIGSjwYGNH83:b94fmzl) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/common/router/displaymanager.service.js#L110)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> Need some comments here explaining what this does.
</details>
Ok added a comment

---

*[src/app/common/router/displaymanager.service.js, line 116 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV-s1DAhnT-vph1vPmD:-KV0CToTb9G7CfkXlBGH:b-gkoyyt) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/common/router/displaymanager.service.js#L116)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> We should really use `storageService` for storing references to panel nodes instead of searching for them everytime we need one.
> 
> Can you use `FOCUS_SELECTORS` in the `find` here?
> 
> Also, why is `$injector` needed?
</details>
Reworked this section

---

*[src/app/layout/focus.service.js, line 40 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV5hkdd9dDv5VmITaiE:-KVQS8BOCcPod8doNBLj:bq9pdyh) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L40)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> I don't think I saw history being assigned to in the code review, if that's the case can you change this to const?
</details>
Changed to `const`

---

*[src/app/layout/focus.service.js, line 77 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV-wRjGZjuKGKLUUjWj:-KVQSZMA4LJVlL3jbWQU:b-o146dh) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L77)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> Maybe this can be moved to the bootstrap as well since this only being executed once
</details>
Moved along with entire focus manager

---

*[src/app/layout/focus.service.js, line 95 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV-wMhKsIhC9luEyGwg:-KV0ESu6Spm8CeAfdlrJ:bhudcf3) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L95)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> Arrow functions?
</details>
Since scope is important for `this`, I left them as normal  functions. If there is a better way with arrow functions let me know

---

*[src/app/layout/focus.service.js, line 191 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV5iQrYWMYgdJGPu5SS:-KVQSl5vP6d9pgdptGKV:b-sf2397) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L191)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> add comment about `applyAsync` and tooltips (also convert this to an API call when this moves out of the viewer code)
</details>
`applyAsync` is no longer needed, tooltip problems must be an artifact of an earlier angular material version.

---

*[src/app/layout/focus.service.js, line 296 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV5jZ6yMfj3Os0GjjWb:-KVQTEA-PoHkEIwUZ7m0:b-xpdm9i) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L296)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> since `lastKnownElement` can be a set rename to something like `lastElementSet`
</details>
Renamed to `lastElementSet`

---

*[src/app/layout/focus.service.js, line 307 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV5jFkd-QkU446mRWna:-KVQVM_mWCQgnND1AzEm:b-rolmkc) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L307)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> add a comment about how this handles "bouncing"
</details>
clarified this line with some comments

---

*[src/app/layout/focus.service.js, line 379 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV5k6hycGvCS94on8p7:-KVQVZHqa2uhiW4-Y3wY:b-q4649j) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L379)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> replace with call to `lastVisibleHistoryElement`
</details>
Replaced with call to `lastVisibleHistoryElement()`

---

*[src/app/layout/focus.service.js, line 405 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV60bA0hLxupt7NUoO0:-KVQVpYnY1gLYlKP5V2A:bu7b836) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L405)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> this function can be inlined into the `else` block below
</details>
inlined this (along with some other minor tweaks)

---

*[src/app/layout/shell.directive.js, line 78 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV63Al_5OKsvdL-TUdJ:-KV_Jyu-kUYh3Ci3wdTh:byqoywv) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/shell.directive.js#L78)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> create a `dialogService` which can wrap `mdDialog.open`
</details>
Created a decorator

---

*[src/app/ui/filters/filters-default-menu.directive.js, line 47 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV63wuO0sxY7hWR99wg:-KVQZ5-EfAsZ9jL6NHPX:b-uchobh) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/ui/filters/filters-default-menu.directive.js#L47)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> there should be an appInfo service which provides the `id` directly instead of looking up every time
</details>
Switched to `appInfo.id`

---

*[src/app/ui/loader/loader-menu.directive.js, line 37 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV64_GRjHlT273UHGk5:-KVQZ9CGn4DO86SGrhud:b-uchobh) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/ui/loader/loader-menu.directive.js#L37)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> appInfo?
</details>
Switched to `appInfo.id`

---

*[src/content/styles/layout/_main.scss, line 55 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV64lxYLsvsK6IA_I_e:-KVQ_42WxXslhTtjXMOK:bo17fpe) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/content/styles/layout/_main.scss#L55)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> add a comment about what this does
</details>
Added a comment to clarify

---

*[src/index-fgp-en.html, line 452 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV-pudoqIHnc5BcOEq5:-KV08DTOMv-FsH0PXe0w:biumyke) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/index-fgp-en.html#L452)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> We should not make the external markup more complicated when not necessary. `rv-trap-focus` attribute can be set the `bootstrap.js`.
</details>
Agreed, moved to `bootstrap.js`

---

*[src/content/styles/modules/_toc.scss, line 320 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV64rbqBq6N9n-lwepm:-KVQ_UR-LnKkEA3hyBjW:b-iwda8k) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/61daf32952e4b1772429b35d47e3d8cf48091bf7/src/content/styles/modules/_toc.scss#L320)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> add this back and figure out a good way to handle this during the next UI meeting
</details>
Added back

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-:-KV_JDF8HBkI0VCwTLm0:b-com43h)*
<!-- Sent from Reviewable.io -->

milesap commented:
Language support for dialog and some documentation additions

---

Review status: 9 of 42 files reviewed at latest revision, 19 unresolved discussions.

---

*[src/app/layout/focus.service.js, line 41 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV5hcC4jPrO5UIGcS7C:-KVaCG2-haQp4qv0Qh31:b-2zksk4) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L41)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> Can you expand the comment to describe the structure of the elements?  Also, is this possible to const?
</details>
I put an example of the `linkedList` structure. There is an assignment on `linkedList` preventing it from being a `const`

---

*[src/app/layout/focus.service.js, line 288 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV5jk6ZSHEU4h7kOevr:-KVaI2ZO3qDMcqrcNUJi:b-dhsdez) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L288)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> add examples in jsdoc: @james-rae can make some fancy mspaint diagrams
</details>
I added a few to get started

---

*[src/app/layout/focus.service.js, line 430 at r1](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KV62oKMcZDnVLVBfuz0:-KVa9AVoVw4v24fV7vJV:bmb1ogc) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0ecd4d8073e97a31f8363e917ba35112aadcd2c5/src/app/layout/focus.service.js#L430)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> translations needed (use substitution strings for the HTML)
</details>
Added translation

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-:-KVaIHqiovgX_f5hf27-:b-jl6kmd)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
Reviewed 3 of 37 files at r1, 13 of 41 files at r2.
Review status: 22 of 42 files reviewed at latest revision, 25 unresolved discussions, some commit checks failed.

---

*[src/app/bootstrap.js, line 176 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVz_l14D3zszq9TyEeS:-KVz_l14D3zszq9TyEeT:blp72ck) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/bootstrap.js#L176)):*
> ```JavaScript
>         }
> 
>         node.setAttribute('rv-trap-focus', appId);
> ```

We need to mark nodes before loading scripts otherwise they can potentially be executed before this point. Would it make sense to move 
```js
scriptsArr.forEach(src => loadScript(src));
 
    // load core.js last and execute any deferred polyfills/patches
    loadScript(`${repo}/core.js`, () => {
        RV._deferredPolyfills.forEach(dp => dp());
        RV.focusManager.bootstrap();
    });
```
just after this loop?

---

*[src/app/focus-manager.js, line 408 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzcNFnF9eTUTq3nYPc:-KVzcNFnF9eTUTq3nYPd:bduffh2) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/focus-manager.js#L408)):*
> ```JavaScript
>      * Initializes document event listeners and contains the jQuery prototypes overrides
>      */
>     function bootstrap() {
> ```

Can we rename this as we already have bootstrap.js.

---

*[src/app/layout/shell.directive.js, line 37 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzdYIGZfE0eHq2-CT0:-KVzdYIGZfE0eHq2-CT1:b-nd1lp6) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/layout/shell.directive.js#L37)):*
> ```JavaScript
>             $rootScope.$on(events.rvApiReady, () => {
>                 $rootElement.find('.rv-esri-map svg').attr('focusable', false);
>                 globalRegistry.focusManager.addViewer($rootElement, $mdDialog);
> ```

This call can be moved to `core.run.js`: https://github.com/fgpv-vpgf/fgpv-vpgf/blob/develop/src/app/core/core.run.js#L168
There is already a call to register a map, so it seems a good place to me.

Also, we have a `rvApiHalt`, which temporarily suspends all external API calls to the viewer, until the next `rvApiReady` event. This will register the same viewer again with the focus manager in such cases. Probably needs a check if the viewer was already added to the focus manager list.

---

*[src/app/ui/basemap/basemap.html, line 1 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVze_1jh8a0w-OvMPuR:-KVze_1kly4-zJdoWGHI:b-vq7log) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/ui/basemap/basemap.html#L1)):*
> ```HTML
> <md-sidenav rv-trap-focus rv-basemap class="rv-basemap-selector md-sidenav-right md-whiteframe-z2" md-component-id="right">
> ```

Basemap selector is rendered inside the viewer container. Does it need `rv-trap-focus`?

---

*[src/app/ui/basemap/basemap.service.js, line 53 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzeo0IOfbnHJAZyDvr:-KVzeo0IOfbnHJAZyDvs:b-5cpjq8) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/ui/basemap/basemap.service.js#L53)):*
> ```JavaScript
>             return $mdSidenav('right')
>                 .open()
>                 .then(() => $('md-sidenav[md-component-id="right"]').focus(true));
> ```

Some comments please :)

---

*[src/app/ui/common/dialog.decorator.js, line 39 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzgDe83DF-uuXxdfPQ:-KVzgDe83DF-uuXxdfPR:b5wtlwy) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/ui/common/dialog.decorator.js#L39)):*
> ```JavaScript
>                     // newly created focus trap tabindex must be -1 as required by focus manager for all focusable elements
>                     element
>                         .find('.md-dialog-focus-trap')
> ```

I don't see this class used anywhere. What is it for?

---

*[src/app/ui/loader/loader-service.html, line 1 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzge8uXfQvGkI02O0y:-KVzge8uXfQvGkI02O0z:b1n7lz7) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/ui/loader/loader-service.html#L1)):*
> ```HTML
> <rv-content-pane rv-trap-focus close-panel="self.closeLoaderService()" floating-header="true">
> ```

This one is also inside the viewer.

---

*[src/app/ui/sidenav/sidenav.html, line 1 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzgiSomKBx7_EJVIt_:-KVzgiSomKBx7_EJVIta:bv1lhhl) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/ui/sidenav/sidenav.html#L1)):*
> ```HTML
> <md-sidenav rv-trap-focus class="site-sidenav md-sidenav-left md-whiteframe-z2" md-component-id="left">
> ```

And this is rendered inside.

---

*[src/app/ui/toc/entry.directive.js, line 50 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzgpf38eyALxYo-DaM:-KVzgpf4gLn1GxDIzbfm:b-l3kfh2) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/ui/toc/entry.directive.js#L50)):*
> ```JavaScript
>             const self = scope.self;
> 
>             self.appID = $rootElement.attr('id');
> ```

AppInfo

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
Review status: 22 of 42 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[src/app/ui/toc/entry.directive.js, line 50 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzgpf38eyALxYo-DaM:-KW3ZceCpRd7T3-ZuTsn:b-50dlir) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/ui/toc/entry.directive.js#L50)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> AppInfo
</details>
By AppInfo, I suggest to use `appInfo` value service to get the viewer instance id.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293)*
<!-- Sent from Reviewable.io -->

alyec commented:
Reviewed 16 of 41 files at r2, 5 of 5 files at r3.
Review status: 40 of 42 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[src/app/focus-manager.js, line 318 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KW3X-khTfkXJKoomog6:-KW3X-khTfkXJKoomog7:b-i4jxer) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/focus-manager.js#L318)):*
> ```JavaScript
>      * @private
>      * @function onFocusin
>      */
> ```

missing arg in jsdoc

---

*[src/app/focus-manager.js, line 357 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KW3XFLPt-UWGaGiMnw3:-KW3XFLPt-UWGaGiMnw4:b-s9ceoi) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/focus-manager.js#L357)):*
> ```JavaScript
>                 if (keys[27]) { // escape key
>                     viewerActive.setStatus(statuses.INACTIVE);
>                 } else {
> ```

`} else { if {` isn't the middle block redundant?

Also, maybe splitting up the key detection from the actions might help with the complexity.

---

*[src/app/focus-manager.js, line 529 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzhbD77MhhBQeDDxbm:-KVzhbD77MhhBQeDDxbn:b-bfwc4n) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/focus-manager.js#L529)):*
> ```JavaScript
>      */
>     function disableCommonPrototypes(funcName) {
>         return (function () {
> ```

Safe to use arrow functions since this will go through babel.

---

*[src/app/focus-manager.js, line 530 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzhYuuAu5M4Pz7pixM:-KVzhYuuAu5M4Pz7pixN:b-7s4sqq) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/focus-manager.js#L530)):*
> ```JavaScript
>     function disableCommonPrototypes(funcName) {
>         return (function () {
>             var originalFunc = jQuery.Event.prototype[funcName];
> ```

const

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293)*
<!-- Sent from Reviewable.io -->

alyec commented:
Reviewed 12 of 41 files at r2.
Review status: 40 of 42 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293)*
<!-- Sent from Reviewable.io -->

milesap commented:
Review status: 33 of 43 files reviewed at latest revision, 15 unresolved discussions.

---

*[src/app/bootstrap.js, line 176 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVz_l14D3zszq9TyEeS:-KW9A1KXDAQqlvwjipKk:bm847y9) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/bootstrap.js#L176)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> We need to mark nodes before loading scripts otherwise they can potentially be executed before this point. Would it make sense to move 
> ```js
> scriptsArr.forEach(src => loadScript(src));
>  
>     // load core.js last and execute any deferred polyfills/patches
>     loadScript(`${repo}/core.js`, () => {
>         RV._deferredPolyfills.forEach(dp => dp());
>         RV.focusManager.bootstrap();
>     });
> ```
> just after this loop?
</details>
Good point, I moved the node tagging ahead of script loading

---

*[src/app/focus-manager.js, line 318 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KW3X-khTfkXJKoomog6:-KW9Al7SmmJxljI4gEeo:b-3xpnio) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/focus-manager.js#L318)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> missing arg in jsdoc
</details>
Added missing jsdoc

---

*[src/app/focus-manager.js, line 357 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KW3XFLPt-UWGaGiMnw3:-KW9ER4wwnddAcwhx20-:b-v5qrbd) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/focus-manager.js#L357)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> `} else { if {` isn't the middle block redundant?
> 
> Also, maybe splitting up the key detection from the actions might help with the complexity.
</details>
I changed the structure slightly so that (hopefully) it's a bit easier to read. Added some additional comments for clarification.

---

*[src/app/focus-manager.js, line 408 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzcNFnF9eTUTq3nYPc:-KW9FUyW06SdBCSBlD2C:b85qk2u) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/focus-manager.js#L408)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> Can we rename this as we already have bootstrap.js.
</details>
Ok, renamed to `init`

---

*[src/app/focus-manager.js, line 529 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzhbD77MhhBQeDDxbm:-KW9ITcTu-0LLku3TMRP:bxqe35d) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/focus-manager.js#L529)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> Safe to use arrow functions since this will go through babel.
</details>
Used arrow function here, (but not in inner function) for scope issue. 

---

*[src/app/focus-manager.js, line 530 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzhYuuAu5M4Pz7pixM:-KW9IR7a7foWSZ0W3MEQ:bq9pdyh) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/focus-manager.js#L530)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> const
</details>
Changed to `const`

---

*[src/app/layout/shell.directive.js, line 37 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzdYIGZfE0eHq2-CT0:-KW9O1SoxolCwiqScWcF:bqszz31) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/layout/shell.directive.js#L37)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> This call can be moved to `core.run.js`: https://github.com/fgpv-vpgf/fgpv-vpgf/blob/develop/src/app/core/core.run.js#L168
> There is already a call to register a map, so it seems a good place to me.
> 
> Also, we have a `rvApiHalt`, which temporarily suspends all external API calls to the viewer, until the next `rvApiReady` event. This will register the same viewer again with the focus manager in such cases. Probably needs a check if the viewer was already added to the focus manager list.
</details>
Moved to `core.run.js`.

Added a check on `addViewer` function so that duplicate viewers are not added to the focus list.

---

*[src/app/ui/basemap/basemap.html, line 1 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVze_1jh8a0w-OvMPuR:-KW9OUnWSeAVh0-MvsS4:b-6xgawg) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/ui/basemap/basemap.html#L1)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> Basemap selector is rendered inside the viewer container. Does it need `rv-trap-focus`?
</details>
It should have `rv-trap-focus` applied so that focus remains inside the basemap selector (and wraps to the top/bottom as needed). This prevents focus from being applied to elements outside the basemap selector that are under the overlay, or hidden by the basemap itself (mapnav buttons).

---

*[src/app/ui/basemap/basemap.service.js, line 53 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzeo0IOfbnHJAZyDvr:-KW9Qcoe4H9Tvj6fu_GD:b-w9u2yg) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/ui/basemap/basemap.service.js#L53)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> Some comments please :)
</details>
Added a comment here to clarify

---

*[src/app/ui/common/dialog.decorator.js, line 39 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzgDe83DF-uuXxdfPQ:-KW9QsvEtMqCaek-4JnO:b-4rd5qq) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/ui/common/dialog.decorator.js#L39)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> I don't see this class used anywhere. What is it for?
</details>
When `md-dialog` is created, a `<div class="md-dialog-focus-trap" tabindex="0"></div>` element is created immediately before and after the `md-dialog`.  These should be tabindex `-1` along with all other focusable elements (or you could argue these elements be removed from the DOM altogether)

---

*[src/app/ui/loader/loader-service.html, line 1 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzge8uXfQvGkI02O0y:-KW9SxInX4vXWS9TtPbp:b-6xnnes) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/ui/loader/loader-service.html#L1)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> This one is also inside the viewer.
</details>
This one is optional if you want focus kept inside the import panel. Let me know if we should keep/remove this trap.

---

*[src/app/ui/sidenav/sidenav.html, line 1 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzgiSomKBx7_EJVIt_:-KW9SpEi0coPfh4fa3LN:b-v66c3w) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/ui/sidenav/sidenav.html#L1)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> And this is rendered inside.
</details>
Kept for same reason as basemap (see above)

---

*[src/app/ui/toc/entry.directive.js, line 50 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzgpf38eyALxYo-DaM:-KW9T_8toj7Z-U37A3ya:b-8j6voa) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/ui/toc/entry.directive.js#L50)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> By AppInfo, I suggest to use `appInfo` value service to get the viewer instance id.
</details>
When I used `appInfo.id` instead of `$rootElement.attr('id')` on `index-many.html` it fails. 

Doing a `console.debug(appInfo.id, $rootElement.attr('id'));` indicates diverging values with `appInfo.id` having a value of `rv-app-2` at all times while `$rootElement.attr('id')` seems to correctly show the actual app id.

Did the same test in `shell.directive.js` however the two values were consistent. If you have any ideas let me know. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293)*
<!-- Sent from Reviewable.io -->

alyec commented:
Focus moved here after opening the side panel:
![2016-11-10 08\_53\_19-title - Firefox Developer Edition.png](https://files.reviewableusercontent.io/images/reviewable/7646414/ohkjtsB3nqubevtr6ILdjlAS/2016-11-10%2008_53_19-title%20-%20Firefox%20Developer%20Edition.png)

`+` and `-` don't zoom the map

Reorder mode `freezes` after moving one position

---

Reviewed 7 of 8 files at r4.
Review status: 40 of 43 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[src/app/focus-manager.js, line 357 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KW3XFLPt-UWGaGiMnw3:-KWDWaioCDuBo_AAsYnw:bhmifpg) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/focus-manager.js#L357)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary>
> I changed the structure slightly so that (hopefully) it's a bit easier to read. Added some additional comments for clarification.
</details>
Thanks, definitely reads better now.

---

*[src/app/core/core.run.js, line 1 at r4](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KWDX2754SFrZ15uqGQH:-KWDX276dU9NCkQVM2B0:b-bxoog7) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/cbe83ad0c6ad661421513ea9ebd2b60012445ae8/src/app/core/core.run.js#L1)):*
> ```JavaScript
> // jshint maxparams:12
> ```

Leave a FIXME so we can figure out what is wrong with this block at some point

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-:-KWDVcI0EkQ_zwcQSnOO:bawnqts)*
<!-- Sent from Reviewable.io -->

milesap commented:
- Removed the blue border around focused links which failed to render in the correct location
- Fixed `+` and `-` issue - FF and IE are using different key codes for those particular keys
- Fixed reorder freezing - custom focus behaviour for dragula was being rejected due to missing `focus` parameter, changed to `focus(true)` 

---

Review status: 36 of 43 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[src/app/core/core.run.js, line 1 at r4](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KWDX2754SFrZ15uqGQH:-KWDyL_LTKePIfZd3z47:b-865l5j) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/cbe83ad0c6ad661421513ea9ebd2b60012445ae8/src/app/core/core.run.js#L1)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary>
> Leave a FIXME so we can figure out what is wrong with this block at some point
</details>
Added a FIXME

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-:-KWDx65SHWMB5gp2vtOB:b-mspggo)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
Reviewed 3 of 41 files at r2, 1 of 8 files at r4.
Review status: 36 of 43 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
Reviewed 5 of 37 files at r1, 7 of 41 files at r2, 3 of 5 files at r3, 5 of 8 files at r4, 3 of 4 files at r5.
Review status: 41 of 43 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[src/app/ui/basemap/basemap.html, line 1 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVze_1jh8a0w-OvMPuR:-KWY8AhAJPdcHQjDgTjw:bxu90ku) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/ui/basemap/basemap.html#L1)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary>
> It should have `rv-trap-focus` applied so that focus remains inside the basemap selector (and wraps to the top/bottom as needed). This prevents focus from being applied to elements outside the basemap selector that are under the overlay, or hidden by the basemap itself (mapnav buttons).
</details>
Awesome, can you add this as a comment here? :)

---

*[src/app/ui/toc/entry.directive.js, line 50 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzgpf38eyALxYo-DaM:-KWYN--3j2Lv0btueb6q:b-1gzqv5) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/ui/toc/entry.directive.js#L50)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary>
> When I used `appInfo.id` instead of `$rootElement.attr('id')` on `index-many.html` it fails. 
> 
> Doing a `console.debug(appInfo.id, $rootElement.attr('id'));` indicates diverging values with `appInfo.id` having a value of `rv-app-2` at all times while `$rootElement.attr('id')` seems to correctly show the actual app id.
> 
> Did the same test in `shell.directive.js` however the two values were consistent. If you have any ideas let me know. 
</details>
There was a problem with the `appInfo` value service. See this PR: https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/1364#-

---

*[src/content/styles/base/_all.scss, line 30 at r5](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KWY5RssCG1_Sctl0qQk:-KWY5RssCG1_Sctl0qQl:b-ryulci) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/02afdf5e93382466c98b5bb58b3f529d2dfe117a/src/content/styles/base/_all.scss#L30)):*
> ```SCSS
>     border-bottom: 1px solid $primary-color-light;
>     transition: border-bottom 0.35s;
>     outline: none !important;
> ```

This removes outline from the link elements when they have focus. We should not be modifying this as it's now impossible to tell if the link is focused or not. Below, focus is set to the "Data Catalogue Page" link, but you can't tell that visually.
![](https://i.imgur.com/XYAJROv.png)

What was this needed for anyway?

---

*[src/content/styles/layout/_main.scss, line 53 at r5](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KWY6CmabU2L-qQwZjL9:-KWY6CmbzO7AUq8Hm_-8:b-khible) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/02afdf5e93382466c98b5bb58b3f529d2dfe117a/src/content/styles/layout/_main.scss#L53)):*
> ```SCSS
>     }
> }
> ```

This only applies to the focus manager dialog, right? If so, can you make a separate class for it, otherwise it's not clear if it applies to anything else in the viewer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
I noticed that skipping the focus dialog and then pressing `Shift + Tab` sometimes sets focus on the viewer directly without triggering the dialog. I think it might be related to that 200 ms delay - if `Shift + Tab` is pressed in that time window.

---

Review status: 41 of 43 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[src/app/ui/loader/loader-menu.html, line 14 at r5](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KWYQjRrA3XhaG2i8ymK:-KWYQjRrA3XhaG2i8ymL:bwovzzj) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/02afdf5e93382466c98b5bb58b3f529d2dfe117a/src/app/ui/loader/loader-menu.html#L14)):*
> ```HTML
>     </md-button>
> 
>     <md-menu-content rv-trap-focus="{{self.appID}}" class="rv-menu rv-dense" width="4">
> ```

Can you also use one-time binding for the app id? It's not supposed to change anyway.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-:-KWYPD1ZmrIQRELyX8Tt:b8cf09z)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
Review status: 41 of 43 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[src/app/ui/common/dialog.decorator.js, line 39 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzgDe83DF-uuXxdfPQ:-KWYS2huZWzHUEoSz314:b-l9ljhu) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/ui/common/dialog.decorator.js#L39)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary>
> When `md-dialog` is created, a `<div class="md-dialog-focus-trap" tabindex="0"></div>` element is created immediately before and after the `md-dialog`.  These should be tabindex `-1` along with all other focusable elements (or you could argue these elements be removed from the DOM altogether)
</details>
Ok, thanks for explanation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293)*
<!-- Sent from Reviewable.io -->

alyec commented:
Reviewed 2 of 41 files at r2, 1 of 8 files at r4, 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293)*
<!-- Sent from Reviewable.io -->

milesap commented:
Waiting status was being set after the dialog opened so that `tabindex=-1` can be applied to dialog elements as well. Moved the logic so that waiting status is updated immediately, with `tabindex` being reset after the dialog opens.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[src/app/ui/basemap/basemap.html, line 1 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVze_1jh8a0w-OvMPuR:-KWZ43vGqMsXHuAIwWoJ:b-cqlbt9) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/ui/basemap/basemap.html#L1)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> Awesome, can you add this as a comment here? :)
</details>
Good idea, added html comment

---

*[src/app/ui/common/dialog.decorator.js, line 39 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzgDe83DF-uuXxdfPQ:-KWZ47vZn0jJlOE6rSRO:b-8fx2u7) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/ui/common/dialog.decorator.js#L39)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> Ok, thanks for explanation.
</details>
:+1: 

---

*[src/app/ui/loader/loader-menu.html, line 14 at r5](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KWYQjRrA3XhaG2i8ymK:-KWZ5H8e17uXR8CMYzb_:b-wexhev) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/02afdf5e93382466c98b5bb58b3f529d2dfe117a/src/app/ui/loader/loader-menu.html#L14)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> Can you also use one-time binding for the app id? It's not supposed to change anyway.
</details>
Changed all app ids to one time bindings

---

*[src/app/ui/toc/entry.directive.js, line 50 at r3](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KVzgpf38eyALxYo-DaM:-KWZ5SB69WU75Dex2_J4:be1iwnl) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/509be16cd03e0d8a83d9eaf136905f2652516187/src/app/ui/toc/entry.directive.js#L50)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> There was a problem with the `appInfo` value service. See this PR: https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/1364#-
</details>
Changed `$rootElement` to `appInfo` per #1364 PR

---

*[src/content/styles/base/_all.scss, line 30 at r5](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KWY5RssCG1_Sctl0qQk:-KWZ6OY6HxGgXHNpc_IO:b-9lerya) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/02afdf5e93382466c98b5bb58b3f529d2dfe117a/src/content/styles/base/_all.scss#L30)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> This removes outline from the link elements when they have focus. We should not be modifying this as it's now impossible to tell if the link is focused or not. Below, focus is set to the "Data Catalogue Page" link, but you can't tell that visually.
> ![](https://i.imgur.com/XYAJROv.png)
> 
> What was this needed for anyway?
</details>
Restored link element outline on focus - originally removed since the border did not wrap the link, it appeared below to the right instead. Fixed css so this does not happen

---

*[src/content/styles/layout/_main.scss, line 53 at r5](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KWY6CmabU2L-qQwZjL9:-KWZ942hiLimL6Pz-KIV:bv1pl7t) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/02afdf5e93382466c98b5bb58b3f529d2dfe117a/src/content/styles/layout/_main.scss#L53)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary>
> This only applies to the focus manager dialog, right? If so, can you make a separate class for it, otherwise it's not clear if it applies to anything else in the viewer.
</details>
Made a `sm-button` class for divs so this is more clear.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-:-KWZ3BDTkOQQ_YImF_eD:b-gzb0ih)*
<!-- Sent from Reviewable.io -->

alyec commented:
Seems like `::after` expands the size of the sidenav:
![2016-11-14 14\_04\_46-title - Waterfox.png](https://files.reviewableusercontent.io/images/reviewable/7646414/U8eIyA8LI1L5AunYwdb87Go8/2016-11-14%2014_04_46-title%20-%20Waterfox.png)
![2016-11-14 14\_05\_49-title - Waterfox.png](https://files.reviewableusercontent.io/images/reviewable/7646414/_NV0WJefkKdPfoWOc6xSZs0d/2016-11-14%2014_05_49-title%20-%20Waterfox.png)

Arrow keys do not work in sidenav or the legend.

Not sure if related or separate issue: when activating import layer with the keyboard focus is not placed in the dialog.

I think splitting out separate issues is okay if we want to merge the PR.

---

Reviewed 15 of 15 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-:-KWZG2xTVXLy_KjyRM6M:bmsv5bq)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
> Waiting status was being set after the dialog opened so that tabindex=-1 can be applied to dialog elements as well. Moved the logic so that waiting status is updated immediately, with tabindex being reset after the dialog opens.

This is still happening :(

---

Reviewed 12 of 15 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/app/ui/loader/loader-menu.html, line 14 at r5](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KWYQjRrA3XhaG2i8ymK:-KWZBWuZ1k2QUuwnhvIZ:b-ucfknh) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/02afdf5e93382466c98b5bb58b3f529d2dfe117a/src/app/ui/loader/loader-menu.html#L14)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary>
> Changed all app ids to one time bindings
</details>
Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-:-KWZL6QKApKRWkpGPPY_:b-mdby63)*
<!-- Sent from Reviewable.io -->

alyec commented:
Sorry, one more issue probably related to the focus highlights.  When elements are highlighted they change sizes (at least some of them, like the image in the sidenav and the links in metadata) and these should probably stay the same size so that we don't reflow.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-:-KWZNemcp3oej6kwx3Wh:b-7ma1w8)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
:point_up: I think this might be related with the change here: https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/1293#-KWY5RssCG1_Sctl0qQk

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-:-KWZOE-TydtwaLXidpdJ:b-he9alg)*
<!-- Sent from Reviewable.io -->

milesap commented:
@aleksueir try it again now - updated when tabindex resets take place

@alyec I changed the way focused elements are shown - no longer using `border` and instead using `outline`. This will not cause reflow. The only issue it seems is that md-sidenav only has a right outline visible since top, left, bottom are positioned absolutely 0px. 

The arrow keys and activating import layer can be opened as separate issues if needed. 

---

Review status: 40 of 43 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-:-KWbyyatZsuEwhWazY1n:b-jt37fl)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
@milesap I still see the skipping behaviour :(  
@milesap @alyec I suggest we merge this PR, and create separate issues to address any further problems.

---

Reviewed 1 of 15 files at r6, 4 of 4 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-:-KWcCZyQV18tk6c5bwTg:b-q6b5k)*
<!-- Sent from Reviewable.io -->

milesap commented:
- changed sidenav panels so that focus is on first button (close button) instead of the panel itself

---

Review status: 37 of 43 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-:-KWcSj7IPcFkkN99NgGo:b-lwmivj)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>
Please create github issues we talked about.

---

Reviewed 6 of 6 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-:-KWcUID3BcQP1qdg9a45:bdcbkms)*
<!-- Sent from Reviewable.io -->

alyec commented:
Reviewed 4 of 4 files at r7, 6 of 6 files at r8.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293)*
<!-- Sent from Reviewable.io -->

alyec commented:
Needs https://github.com/fgpv-vpgf/fgpv-vpgf/pull/1364 merged as a dependency

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-:-KWcWpiG6ClpdqFg2qXL:bx0jbng)*
<!-- Sent from Reviewable.io -->

milesap commented:
#1364 has been merged

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-:-KWcZyYOAZyzfY2R2B7X:b-iryl0o)*
<!-- Sent from Reviewable.io -->

alyec commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (basic testing in Firefox)

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-:-KWcajg5tXz2yodZlSDk:b-z7r68)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> in Chrome.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1293#-:-KWcanY3UIjY49emAxAl:b7epci0)*
<!-- Sent from Reviewable.io -->
